File: C#/Stack/Stack.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Stack
{
	public class StackNode<T> 
	{
		private class StackNode<T>
		{
			public T Data { get; set; }
			public StackNode<T> Next { get; set; }
			
			public StackNode(T e) 
			{
				Data = e;
				Next = null;
			}
		}
		
		private StackNode<T> _top;
		
		public int Count { get; }
		
		public Stack() 
		{
			_top = null;
			Count = 0;
		}
		
		public Stack(T[] elems) : this() 
		{
			foreach (T e in elems)
			{
				Push(e);
			}
		}
		
		public bool IsEmpty() 
		{
			return Count == 0;
		}
		
		public void Push(T e) 
		{
			if (Count == 0) 
			{
				_top = new StackNode<T>(e);
			}
			else 
			{
				StackNode<T> temp = new StackNode<T>(e);
				temp.Next = _top;
				_top = temp;
			}
			Count++;
		}
		
		public T Peek()
		{
			return _top?.Data;
		}
		
		public T Pop()
		{
			if (Count == 0) 
			{
				return null;
			}
			T data = _top.Data;
			_top = _top.Next;
			Count--;
			return data;
		}
		
		public bool Contains(T e) 
		{
			StackNode<T> current = _top;
			while (current != null)
			{
				if (current.Data.Equals(e))
				{
					return true;
				}
				current = current.Next;
			}
			return false;
		}
		
		public void Clear() 
		{
			_top = null;
			Count = 0;
		}
	}
}

File: Java/stack/Stack.java
package stack;

public class Stack<E> {
	private class StackNode<E> {
		private E data;
		private StackNode<E> next;
	
		private StackNode(E e) {
			data = e;
			next = null;
		}
	}
	
	private StackNode<E> top;
	private int size;
	
	public Stack() {
		top = null;
		size = 0;
	}
	
	public Stack(E[] elems) {
		this();
		for (E e : elems) {
			push(e);
		}
	}
	
	public int size() {
		return size;
	}
	
	public boolean isEmpty() {
		return size == 0;
	}
	
	public void push(E e) {
		if (size == 0) {
			top = new StackNode<E>(e);
		}
		else {
			StackNode<E> temp = new SinglyNode<E>(e);
			temp.next = top;
			top = temp;
		}
		size++;
	}
	
	public E peek() {
		if (top == null) {
			return null;
		}
		return top.data;
	}
	
	public E pop() {
		if (size == 0) {
			return null;
		}
		E data = top.data;
		top = top.next;
		size--;
		return data;
	}
	
	public boolean contains(E e) {
		StackNode<E> current = top;
		while (current != null) {
			if (current.data.equals(e)) {
				return true;
			}
			current = current.next;
		}
		return false;
	}
	
	public void clear() {
		top = null;
		size = 0;
	}
}
